### DISCUSSION TITLE
Introduce IJobExecutionContextFactory

### DISCUSSION THREAD

**Discussion by drieseng** [quartznet/quartznet#1770] [Ideas]
We currently "new" the **JobExecutionContextImpl** in `JobRunShell.Initialize(QuartzScheduler sched, CancellationToken cancellationToken = default)`, hereby providing no easy way to create a custom **ICancellableJobExecutionContext** instance.

If one would like to have a specialized **ICancellableJobExecutionContext**, then one would have to:
* Introduce a specialized **JobRunShell** class that overrides `Initialize(QuartzScheduler sched, CancellationToken cancellationToken = default)`.
* Introduce a specialized **IJobRunShellFactory** in which `CreateJobRunShell(TriggerFiredBundle bndle)` would return an instance of the specialized **JobRunShell** class.

I propose to introduce an **IJobExecutionContextFactory** which would only have a `ICancellableJobExecutionContext Create(IScheduler scheduler, TriggerFiredBundle firedBundle, IJob job)` method. The **IJobExecutionContextFactory** would either be "injected" into:
* the concrete **IJobRunShellFactory** instance, which in turn would pass it on to the **JobRunShell** via its ctor.
* the **QuartzSchedulerThread** ctor which in turn would pass it on to the **JobRunShell** via the `JobRunShell.Initialize(...)` method.

We can use this to have a specialized **IJobExecutionContextFactory** in **Quartz.Extensions.DependencyInjection** that would produce **WrappedJobExecutionContext(Impl)** instances. This **WrappedJobExecutionContext(Impl** class would derive from **JobExecutionContextImpl** and override the **JobInstance** property to return the underlying job.  This in turn would allow to remove the "leakage" (and with it the minor impact on performance) of Microsoft DI in **JobExecutionContextImpl**, and allow us to move **IJobWrapper** to the **Quartz.Extensions.DependencyInjection** assembly.

I'm still thinking this through, but feel free to comment already.

A few other questions:
* Can we make **StdJobRunShellFactory** sealed? If you want to override any of the two methods, you can just as well created a custom **IJobRunShellFactory**.
* Why doesn't `IJobRunShellFactory.CreateJobRunShell(TriggerFiredBundle bndle)` return an **IJobRunShell** instance.

**Note:**
If you do not reject the idea immediately, I'd be happy to implement it in a PR knowing that it may still be rejected after all.

**Comment by lahma** [comment]
This might also be related https://github.com/quartznet/quartznet/discussions/1348 , so maybe custom job run shell is needed?

> Can we make StdJobRunShellFactory sealed? If you want to override any of the two methods, you can just as well created a custom IJobRunShellFactory.

We can seal as much as we want in v4, we then un-seal when there's an outcry...

> Why doesn't IJobRunShellFactory.CreateJobRunShell(TriggerFiredBundle bndle) return an IJobRunShell instance.

Probably Java API decision?

So I wouldn't want to introduce too many new concepts, which might increase the .NET/Java version disparity further. Ideally the concepts would work between the languages with minor runtime-specific twists, having new interfaces and concepts might make this harder.
